libvast/vast/concept/parseable/core/parser.hpp

```diff
@@ -57,35 +57,36 @@ struct parser_base {
     return derived().parse(f, l, a) && f == l;
   }
 
-  // FIXME: don't ignore ADL.
-  template <class Range, class Attribute = unused_type>
-  auto operator()(Range&& r, Attribute& a = unused) const
-  -> decltype(std::begin(r), std::end(r), bool()) {
-    auto f = std::begin(r);
-    auto l = std::end(r);
-    return derived().parse(f, l, a) && f == l;
-  }
-
-  // FIXME: don't ignore ADL.
-  template <class Range, class A0, class A1, class... As>
-  auto operator()(Range&& r, A0& a0, A1& a1, As&... as) const
-  -> decltype(std::begin(r), std::end(r), bool()) {
-    auto t = std::tie(a0, a1, as...);
-    return operator()(r, t);
-  }
-
-  template <class Iterator, class Attribute>
-  auto operator()(Iterator& f, const Iterator& l, Attribute& a) const
-  -> decltype(*f, ++f, f == l, bool()) {
-    return derived().parse(f, l, a);
+  template <class... TAttributes>
+  auto operator()(concepts::range auto&& r, TAttributes&... attributes) const {
+    if constexpr (sizeof...(TAttributes) == 0) {
+      auto f = std::begin(r);
+      auto l = std::end(r);
+      return derived().parse(f, l, unused) && f == l;
+    } else if constexpr (sizeof...(TAttributes) == 1) {
+      auto f = std::begin(r);
+      auto l = std::end(r);
+      return derived().parse(f, l, attributes...) && f == l;
+    } else {
+      auto t = std::tie(attributes...);
+      return operator()(r, t);
```

Comment:
Would it make sense here to go for symmetry and to write the last case like this?

```
auto f = std::begin(r);
auto l = std::end(r);
auto t = std::tie(attributes...);
return derived().parse(f, l, t);
```

Or can it happen that `operator()(r, t)` calls into a different function here?

Reply:
That should be safe. It makes the declarations of `f` and `l` common between all the branches, so it's less code in the end too. Just changed 👍